recipe/meta.yaml

```diff
@@ -220,28 +225,31 @@ outputs:
         - $RPATH/ld64.so.2  # [ppc64le]
     requirements:
       build:
-        - {{ compiler('m2w64_c') }}    # [win]
-        - {{ compiler('m2w64_cxx') }}  # [win]
-        - {{ compiler('c') }}
-        - {{ compiler('cxx') }}
-        - {{ stdlib('c') }}
-        - llvm-openmp                  # [osx]
-        - libgomp                      # [linux]
+        - {{ compiler('m2w64_c') }}        # [win]
+        - {{ compiler('m2w64_cxx') }}      # [win]
+        - {{ compiler('m2w64_fortran') }}  # [win]
+        - {{ stdlib('m2w64_c') }}          # [win]
+        - {{ compiler('c') }}              # [unix]
+        - {{ compiler('cxx') }}            # [unix]
+        - {{ stdlib('c') }}                # [unix]
+        - llvm-openmp                      # [osx]
+        - libgomp                          # [not osx]
+        - m2-bash                          # [win]
         - git
-        - make                         # [not win]
-        - posix                        # [win]
+        - make
         - cmake
         - ninja
-        - cross-r-base {{ r_base }}    # [build_platform != target_platform]
-        - r-base                       # [build_platform != target_platform]
-        - r-matrix                     # [build_platform != target_platform]
-        - r-data.table                 # [build_platform != target_platform]
-        - r-magrittr                   # [build_platform != target_platform]
-        - r-jsonlite                   # [build_platform != target_platform]
-        - r-knitr                      # [build_platform != target_platform]
+        - cross-r-base {{ r_base }}        # [build_platform != target_platform]
+        - r-base                           # [build_platform != target_platform]
+        - r-matrix                         # [build_platform != target_platform]
+        - r-data.table                     # [build_platform != target_platform]
+        - r-magrittr                       # [build_platform != target_platform]
+        - r-jsonlite                       # [build_platform != target_platform]
+        - r-knitr                          # [build_platform != target_platform]
       host:
         - {{ pin_subpackage('_r-xgboost-mutex', exact=True) }}
         - {{ pin_subpackage('libxgboost', max_pin='x.x.x') }}
+        - libgomp                          # [not osx]
```

Comment:
This is already added via `libgomp`'s `run_exports/strong` and it is in the top-level. So let's drop this

```suggestion
```

Reply:
If `libgomp` is in `build` requirements, will that still give a `run_export`?  I thought that only happened with `host`, and the top level `libgomp` is in `build`, not `host`.

Reply:
Yes, adding a package in `requirements/build` will ensure its `run_exports/strong` requirements will be applied in `requirements/host` and `requirements/run`. From the [conda-build docs]( 
https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#export-runtime-requirements ):

> With `run_exports` (new in conda-build 3) such a runtime requirement can be implicitly added by host requirements (e.g. libpng exports libpng), and with `run_exports/strong` even by build requirements (e.g. GCC exports libgcc).

`libgomp` specifically adds [`run_exports/strong` on `_openmp_mutex `]( https://github.com/conda-forge/ctng-compilers-feedstock/blob/09c04e6458ab4d79193b4c20a6330e84540108ec/recipe/meta.yaml#L507-L509 ), which ensures `_openmp_mutex ` is automatically added to `requirements/host` and `requirements/run`. `_openmp_mutex` [depends on `libgomp` at runtime ]( https://github.com/conda-forge/ctng-compilers-feedstock/blob/09c04e6458ab4d79193b4c20a6330e84540108ec/recipe/meta.yaml#L579-L581 ). Meaning anywhere `_openmp_mutex` is installed `libgomp` is too, which means anything that had `libgomp` in `requirements/build`. The `_openmp_mutex` also [tests the `libgomp` SO is available and usable]( https://github.com/conda-forge/ctng-compilers-feedstock/blob/09c04e6458ab4d79193b4c20a6330e84540108ec/recipe/meta.yaml#L586-L589 )

Hence [the recommended usage of OpenMP in conda-forge]( https://conda-forge.org/docs/maintainer/knowledge_base/#openmp ) is to only added OpenMP dependencies in `requirements/build` and let `run_exports/strong` of these dependencies to fill in the rest of `requirements` automatically

Should add XGBoost specifically [tests that all linkages are satisfied]( https://github.com/conda-forge/xgboost-feedstock/blob/0cd682cc56d8b5c44b157571fe9587028c0c94a1/conda-forge.yml#L7-L8 ). So if we linked to OpenMP and the needed libraries are not present, the build would error. Note the builds pass following conda-forge OpenMP recommendations meaning this was working as expected

Reply:
Thanks for the detailed explanation.  We had some issues with a few R packages where `libgomp` wasn't being included in `run_exports` and failed linkage checks on Windows.  I need to revisit those with this information - it's possible we were actually hitting a bug with `run_export` propagation rather than simply missing a dependency.  I'll follow up on this on Zulip if the behavior of `libgomp` is inconsistent with what is expected for `run_exports/strong`.

Reply:
Meaning this list: https://github.com/conda-forge/r-base-feedstock/issues/325#issuecomment-2342479080 ?

Scanned through that list and left a few comments in associated PRs and left a few notes

This one ( https://github.com/conda-forge/r-idefix-feedstock/pull/1 ) is notable as it wasn't doing the right thing with OpenMP to start with. Think fixing that would be the best place to focus effort

The others are pretty simple cleanups after that (mainly deleting the extra OpenMP lines)

Reply:
Yes, that list was not all of them but it did cover many.  I'm now realizing the issue might actually have been that `libgomp` had been a Linux-only dependency and needed to be `not osx` instead.  I'll try to change those to just have `libgomp` where it's needed.